torchx/schedulers/ray/ray_driver.py

```diff
@@ -77,77 +111,184 @@ def load_actor_json(filename: str) -> List[RayActor]:
+        """Creating pending tasks to initialize command actors in placement groups"""
+        pg_ix_of_replica: List[int] = [
+            max(0, i - self.min_nnodes + 1) for i in range(len(self.replicas))
+        ]  # the placement group indices for corresponding replicas
+        for i in range(len(self.replicas)):
+            pg_ix = pg_ix_of_replica[i]
+            pg = self.placement_groups[pg_ix]  # find the created placement group
+            replica = self.replicas[i]
+            self.create_and_schedule_actor(pg, replica)
+
+    def run(self) -> None:
+        result: Optional[PlacementGroup]  # execution result
+        need_more_actors: bool = True  # if need more actors
+        command_actors_count: int = 0  # number of created command actors
+        # Await return result of remote ray function and initialize new command actors
+        while len(self.active_tasks) > 0:
+            _logger.info(f"running ray.wait on {self.active_tasks}")
+            # ray.wait is partial waiting
+            # pyre-fixme[16]: Module `worker` has no attribute `wait`.
+            completed_tasks, self.active_tasks = ray.wait(self.active_tasks)
+            # If a failure occurs the ObjectRef will be marked as completed.
+            # Calling ray.get will expose the failure as a RayActorError.
+            for object_ref in completed_tasks:
+                try:
+                    result = ray.get(object_ref)  # pyre-ignore
+                    if isinstance(result, CommandActorScheduled) and need_more_actors:
+                        actor = self.actor_info_of_id[result.id].actor
+                        if self.master_node_id is None:
+                            # make this actor be the master node
+                            self.master_node_id = result.id
+                            self.rank_0_address, self.rank_0_port = ray.get(
+                                actor.get_actor_address_and_port.remote()
+                            )
+                            self.active_tasks.append(
+                                actor.exec_module.remote("localhost", 0, result.id)
+                            )
+                        else:
+                            self.active_tasks.append(
+                                actor.exec_module.remote(
+                                    self.rank_0_address, self.rank_0_port, result.id
+                                )
+                            )
+                        command_actors_count += 1
+                    elif isinstance(result, TaskCompleted):
+                        need_more_actors = False  # don't need more actors
+                        command_actors_count -= 1  # 1 completed command actor
+                        self.pop_actor_info(result.id)
+                        if (
+                            command_actors_count == 0
+                        ):  # all the command actors have finished
+                            break  # exit
```

Comment:
this only breaks the inner loop -- is there a case where active_tasks >0 but command_actors_count == 0?

Reply:
Yes, active_tasks may contain some actors which haven't been scheduled successfully in a placement group, even after the job has finished. I guess I have to use return here.